documentation/docs/advanced/storage_interface.mdx

```diff
@@ -0,0 +1,106 @@
+
+## Indexing
+
+The interface has two methods called `did_list` and `did_exists`. These return the list of stored DIDs, and whether a DID exists in storage, respectively. Implementations are thus expected to maintain a list or index of stored DIDs. An identity created with `did_create` is added to the index, while an identity deleted through `did_purge` is removed from the index.
+
+If the storage implementation can be accessed concurrently, then access to the index needs to be synchronized, since it is unique per storage instance.
+
+## Implementation
+
+The identity.rs framework ships two implementations of `Storage`. The `MemStore` is an insecure in-memory implementation intended as an example implementation and for testing. The secure and recommended `Storage` is `Stronghold`, however `Stronghold` may also be interesting for implementers to look at, as it needs to deal with some challenges the in-memory version does not have.
+
+This section will detail some common challenges and embeds the `MemStore` implementations in Rust and TypeScript.
+
+### Challenges
+
+The `did_create` method takes the fragment of the initial verification method, the name of a network in which the DID will eventually exist, and an optional private key. From these inputs, it either generates a key or uses the passed private key to calculate the public key and from that derive the DID. In case a key needs to be generated, the challenge is to obtain the location for the key to be stored at. Since the key location depends on the public key, but key generation likely needs a location for the key to be stored at, there is a circular dependency that needs to be resolved. This can be resolved in at least two ways.
+
+1. Generate the key at a random location, then derive the actual location and move the key there
+2. If moving a key is not possible, then an additional mapping from key location to some storage-internal location identifier can be maintained. Then it's possible to generate the key at some storage-internal location, calculate the key location and store the mapping.
+
+Since this also needs to happen before the DID can be derived from the public key, similar approaches can be used to work around the not-yet available DID partition key. Storages may choose to have one statically identified partition where keys are generated initially, and then moved from there. Storages whose restrictions do not allow for this, may want to use the flattened storage layout described in [example layout](#example-layout) and use the mapping approach.
+
+### Storage Test Suite
+
+The `StorageTestSuite` can be used to test the basic functionality of storage implementations. See its documentation for more details.
```

Comment:
Can we link or does that require Rust API online? Can we already link JS?

Reply:
JS yes, but Rust is not on docs.rs yet, so no place to link to, unfortunately.

Reply:
Ah, actually the test suite is a post 0.5 feature, so I will have to remove this section here and add it later for the `dev` or 0.6 docs.